### DISCUSSION TITLE
Can I generate and download automatically OpenAPI json locally when run the app?

### DISCUSSION THREAD

**Discussion by arthuravila26** [fastapi/fastapi#6877] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:

G'day everyone!

I'm wondering whether is possible to download the OpenAPI json document at the time I start my application. I know that if I run it and go to `https://localhost:8000/docs` I can download it. But my purpose is download it automatically when I run the application, it generate the OpenAPI json and download it to my project.

I already have this:

```
def custom_openapi():
    if app.openapi_schema:
        return app.openapi_schema

    openapi_schema = get_openapi(
        title="Dashboards ",
        version="v1",
        description="Api Dashboard",
        routes=app.routes,
    )
    openapi_schema["info"]["x-logo"] = {
        "url": "https://fastapi.tiangolo.com/img/logo-margin/logo-teal.png"
    }
    app.openapi_schema = openapi_schema
    return app.openapi_schema


app.openapi = custom_openapi


@app.get("/dashboards")
def first():
    logger.info("Worked")
    return {'status': 'OK', 'message': 'Worked just fine!'}
```

When I run it locally, I can get it on /docs just fine. But I must find an way that I also generate it locally at the time that I run the application.

**Comment by Kludex** [comment]
Feel free to improve it:

```python
import json

from aiofile import async_open
from fastapi import FastAPI, Request

app = FastAPI()

@app.get("/")
def potato():
    return None

@app.on_event("startup")
async def startup():
    async with async_open("/tmp/hello.json", 'w+') as afp:
        await afp.write(str(json.dumps(app.openapi()).encode("utf-8")))
```

EDIT: Check https://github.com/tiangolo/fastapi/issues/2877#issuecomment-1314267430

**Comment by arthuravila26** [comment]
Thanks mate!! Helped a lot here!!

**Comment by tiangolo** [comment]
Thanks for the help here @Kludex ! 👏  🙇 

Thanks for reporting back and closing the issue @arthuravila26  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.

**Comment by Kludex** [comment, marked as answer]
My solution is old, use `anyio` instead of `aiofiles`: https://anyio.readthedocs.io/en/stable/streams.html#file-streams

**Comment by mavwolverine** [comment]
Sorry to comment on an old thread, but found an alternative way (would be useful in cicd)

`python -c 'from main import app; print(app.openapi());'`

**Comment by a9a4k** [comment]
Building on top of @Kludex, the new way of doing things

```
from anyio.streams.file import FileWriteStream

@asynccontextmanager
async def lifespan(app: FastAPI) -> AsyncGenerator[None, None]:
    # On Startup

    path = "./openapi.json"
    async with await FileWriteStream.from_path(path) as stream:
        await stream.send(json.dumps(app.openapi()).encode("utf-8"))

    yield

    # On Shutdown
app = FastAPI(lifespan=lifespan)
```
